torch/_torch_docs.py

```diff
@@ -3594,6 +3594,51 @@ def merge_dicts(*dicts):
     RuntimeError: bool value of Tensor with no values is ambiguous
 """.format(**common_args))
 
+add_docstr(torch.kron,
+           r"""
+kron(input, other, *, out=None) -> Tensor
+
+Computes the Kronecker product, denoted by :math:`\otimes`, of :attr:`input` and :attr:`other`.
+
+If :attr:`input` is a :math:`(m \times n)` tensor and :attr:`other` is a
```

Comment:
We didn't discuss this previously, but is the Kronecker product defined if either of A or B aren't matrices?

Reply:
I think the Kronecker product is defined mathematically only for matrices. We can think of vectors as m×1 matrix, and scalars as 1×1 matrix, then everything works.
Vectors are tested [here](https://github.com/pytorch/pytorch/pull/45358/files#diff-7e17421f32124016eb8de04dc2f445da5786a28355e1addc72b305466f590180R202-R204) (as `(4, )` shape).
As for n-dimensional arrays with n>2, NumPy extends the definition [as described in the notes section](https://numpy.org/doc/stable/reference/generated/numpy.kron.html) to "blocks of the second tensor scaled by the first tensor".
So kron does not support batching.

Sometimes it's said that for matrices Kronecker product == tensor outer product, but this is not true for tensors in general. For [the example from Wiki about the tensor product](https://en.wikipedia.org/wiki/Outer_product#The_outer_product_of_tensors), kron would give a tensor with dimensions (3*1, 5*10, 7*100).

Reply:
Being like NumPy seems OK (our goal is to be compatible, after all). Would you add a note about how input and other are treated if they're not matrices? The current docs deal with them as if they must be matrices. Something like:

- NOTE
- The Kronecker product is typically defined only for two matrices
- When either is a scalar or vector it's unsqueezed as...
- When either is an input is a tensor with 3+ dimensions then...

ALTERNATIVELY you could expand the description of the function describe the matrix case, say that scalars and vectors are unsqueezed to be matrices, and THEN define the "general" case. That seems like a more challenging but better approach.

What are your thoughts, @IvanYashchuk?

Reply:
What about in the main description present the general definition first and then mention what it does for matrices.

Computes the Kronecker product of `input` and `other`.
For general n-dimensional tensors this function computes:
... math expression
<a href="https://imgbb.com/"><img src="https://i.ibb.co/Jm3ghVS/image.png" alt="image" border="0"></a><br />

The number of dimensions of `input` and `other` is assumed to be the same and if necessary the smaller tensor is unsqueezed as the larger one.

If `input` is a `(m \times n)` ~~tensor~~ matrix and `other` is a `(p \times q)` ~~tensor~~ matrix, the result will be a `(p*m \times q*n)` block ~~tensor~~ matrix:
... kron definition for matrices from wiki
Scalar and vector inputs are unsqueezed to be matrices.

Reply:
That approach sounds good. Instead of 

"The number of dimensions of input and other is assumed to be the same and if necessary the smaller tensor is unsqueezed as the larger one."

I think it can say "If one tensor has fewer dimensions than the other it is unsqueezed until it has the same number of dimensions." 

That change would make the explicit reference to scalar and vector inputs at the end redundant, so it can be removed.

The last caveat is that this needs to define the dot operator in both equations above.

Reply:
Okay, dot was for the normal multiplication of scalars. Is asterisk (*) preferred?

Reply:
Dot's OK if the documentation defines it (the dot operator is used for so many mathematical operations it's highly ambiguous), an asterisk would also be fine and probably doesn't require definition (it's typically used for elementwise multiplication and scalar multiplication).

Reply:
I've updated the documentation so that the main text describes now the computation for general n-dimensional tensors. I left the matrix case description as a note section.
Here is the image of rendered docs:
<a href="https://ibb.co/WyMZ3Qr"><img src="https://i.ibb.co/C7Ygtqj/image.png" alt="image" border="0"></a>